### ISSUE TITLE
TQ2NZ conversion without imatrix produces degenerate output (repeating single token)

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#1623]
## Summary

`gguf_to_onebp --tq2nz` without `--imatrix` produces models whose output degenerates to a repeated single token — on every model tested. The 2.5 bpw speed win is real (packed TQ2NZ path: Llama-3.2-1B 34.4 → 12.6 ms/token, 2.7×), but the output is unusable.

## Repro

```
./build/gguf_to_onebp models/Llama-3.2-1B-Instruct.Q8_0.gguf /tmp/llama32-1b-instruct-tq2nz.1bp --tq2nz
./build/bench_hip_1bp /tmp/llama32-1b-instruct-tq2nz.1bp 8 2
# Tokens: 14322 3189 3189 3189 3189 3189 3189 3189   <- repeats token 3189
```

Same on Qwen3-0.6B (`105534 131138 70439 88185 246 246 246 246`).

## Analysis

Greedy 2-bit rounding without importance-matrix calibration destroys the weights (all-zero scale groups → repeated degenerate logits). The converter already has `load_imatrix_dat` wired (`--imatrix file.dat`, fixes #1355, entry-count bounds in place) — it just needs a real calibration pass feeding it (llama.cpp-style imatrix or a HF calibration corpus).

## Ask

- Wire an imatrix/calibration path into the conversion pipeline (documented command), or
- Add a quality gate: refuse `--tq2nz` without `--imatrix` until calibration is available.

**Comment by bong-water-water-bong** [comment]
**Update — new evidence: this is likely a correctness defect, not (only) a calibration gap.**

Discriminator test (same GGUF source, three formats):

| Format | Output (first 8 tokens) |
|---|---|
| TQ2NZ bf16 | `14322 3189 3189 3189 3189 3189 3189 3189` |
| TQ2NZ_E4M3 | `60220 82123 52107 82123 82123 82123 82123 82123` |
| TQ1 | generate fails (token -1) — unsupported packed path, expected |

Every 2-bit variant hard-repeats a single token on **every model tested** (Llama-3.2-1B, Llama-3.2-1B-Instruct, Qwen3-0.6B — different stuck tokens: 3189 / 246 / 82123). A genuine quantization-quality issue (even without imatrix) produces diverse-but-imperfect output; a hard single-token repeat across formats and models indicates the 2-bit codebook path — converter rounding or the packed decode kernels (`h1bp_tq2nz_part_kernel` reading tile codes) — is systematically wrong.

Suggested next step: dump the packed logits (`H1BP_DUMP` doesn't fire in graph mode — needs a `--no-graph`-style toggle or eager run) and compare the Q4NX vs TQ2NZ dequantized weights for a known tensor to see where the codes/scales diverge.

**Comment by bong-water-water-bong** [comment]
**Definitive diagnosis — retracting the "codebook defect" theory (comment 5290351639).**

That theory was built on a broken Python parser (tile slice misaligned by 4 bytes vs the loader's ground-truth index). With the engine's OWN loader dequant (`NpuOnebpModel::get_tensor_f32` → `dequant_matrix`) as reference, verified against the source GGUF (`blk.0.attn_q.weight`, Q8_0 → dequantized):

| Quant | corr vs source | SNR |
|---|---|---|
| Q4NX | 0.9967 | **21.8 dB** |
| TQ2NZ | 0.9088 | **7.0 dB** |

Also verified the engine math is correct: L0-Q GEMV output matches `W_dequant @ x` exactly (corr 1.0); the layer dumps show L0-Q at corr 0.996 vs Q4NX — the packed GEMV reads the tiles correctly.

**Verdict: the converter and kernels are correct. The hard-repeat (`14322 3189 3189...`) is inherent 2-bit quantization quality on a 1B model — 7 dB weight SNR is at the 2-bit information limit, and without imatrix the error accumulates across 16 layers and collapses the logits to a fixed token. Same mechanism on Qwen3-0.6B (smaller, worse).** Both bf16 and E4M3 scale variants sit at the same ~7 dB — consistent with a quality limit, not an encoding bug.

The fix direction stands: **imatrix/calibration** (`load_imatrix_dat` is wired, `--imatrix` flag exists) or a quality gate. Diagnostic tooling for this is now in place: #1626 added `H1BP_GRAPH=0` + `H1BP_DUMP` eager-mode dumps (committed 9735974d), which made this whole investigation possible.

**Comment by bong-water-water-bong** [comment]
FINAL RESOLUTION — 2-bit is not viable for this model, even with a real imatrix.

Generated a genuine importance matrix (llama-imatrix, 5k-token calibration corpus on Llama-3.2-1B-Instruct.Q8_0.gguf, 112 tensors) and converted with the wired-in --imatrix path. The E4M3 MSE scale search uses the importance weighting; verified the converter loaded it ('imatrix: 112 tensors loaded').

Results (bench_hip_1bp, 8 tok):
- TQ2NZ bf16 + imatrix: IGNORED (the bf16 path doesn't use importance — converter gap) -> 14322 3189 3189...
- TQ2NZ_E4M3 + imatrix: 90 220 220 220 482 220... — changed quantization, still a hard repeat

Even with importance weighting, 2-bit on a 1B model collapses. The practical formats for this engine are Q4NX (21.8 dB — now runs PACKED at 11.6 ms/token via #1625) or TQ2BS (block-scaled ternary, 5 B/block). Recommendation: ship Q4NX; drop TQ2NZ for sub-3B models. Closing as investigated-to-completion.
